include/ginkgo/core/matrix/row_gatherer.hpp

```diff
@@ -0,0 +1,181 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_MATRIX_ROW_GATHERER_HPP_
+#define GKO_PUBLIC_CORE_MATRIX_ROW_GATHERER_HPP_
+
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <vector>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+/**
+ * RowGatherer is a matrix "format" which stores the row rowgatherer
+ * arrays which can be used to perform rowgatherer rows to another matrix.
+ *
+ * @tparam IndexType  precision of rowgatherer array indices.
+ *
+ * @note This format is used mainly to allow for an abstraction of the
+ * rowgatherer and provides the user with an apply method which
+ * calls the respective LinOp's rowgatherer operation if the respective LinOp
+ * implements the RowGatherable interface. As such it only stores an array of
+ * the rowgatherer indices.
+ *
+ * @ingroup rowgatherer
+ * @ingroup matrix
+ * @ingroup LinOp
+ */
+template <typename IndexType = int32>
+class RowGatherer : public EnableLinOp<RowGatherer<IndexType>>,
+                    public EnableCreateMethod<RowGatherer<IndexType>> {
+    friend class EnableCreateMethod<RowGatherer>;
+    friend class EnablePolymorphicObject<RowGatherer, LinOp>;
+
+public:
+    using index_type = IndexType;
+
+    /**
+     * Returns a pointer to the array of row_gather_index.
+     *
+     * @return the pointer to the row_gather_index array.
+     */
+    index_type* get_row_gather_index() noexcept
```

Comment:
Shouldn't this be plural: `get_row_gatherer_indices`? Same for the function below.